src/sources/file_descriptors/mod.rs

```diff
@@ -0,0 +1,138 @@
+use std::{io, thread};
+
+use async_stream::stream;
+use bytes::Bytes;
+use chrono::Utc;
+use codecs::{
+    decoding::{DeserializerConfig, FramingConfig},
+    StreamDecodingError,
+};
+use futures::{channel::mpsc, executor, SinkExt, StreamExt};
+use tokio_util::{codec::FramedRead, io::StreamReader};
+use vector_core::config::LogNamespace;
+use vector_core::ByteSizeOf;
+
+use crate::{
+    codecs::DecodingConfig,
+    config::log_schema,
+    internal_events::{BytesReceived, OldEventsReceived, StreamClosedError},
+    shutdown::ShutdownSignal,
+    SourceSender,
+};
+
+#[cfg(all(unix, feature = "sources-pipe"))]
+pub mod pipe;
+#[cfg(feature = "sources-stdin")]
+pub mod stdin;
+
+pub trait FileDescriptorConfig {
+    fn host_key(&self) -> Option<String>;
+    fn framing(&self) -> Option<FramingConfig>;
+    fn decoding(&self) -> DeserializerConfig;
+    fn description(&self) -> String;
+}
+
+pub fn file_descriptor_source<R, C>(
+    mut reader: R,
+    config: C,
+    shutdown: ShutdownSignal,
+    mut out: SourceSender,
+    name: &'static str,
+) -> crate::Result<crate::sources::Source>
+where
+    R: Send + io::BufRead + 'static,
+    C: FileDescriptorConfig,
```

Comment:
Consider making this a method of `trait FileDescriptorConfig`:
```suggestion
pub trait FileDescriptorConfig {
    fn source(
        self,
        mut reader: impl Send + io::BufRead + 'static,
        shutdown: ShutdownSignal,
        mut out: SourceSender,
        name: &'static str,
    ) -> crate::Result<crate::sources::Source> {
        …
    }
}
```